Refactor configuration

spacecowboy opened the pull request:
The primary goal of this change is to make it possible to validate a
configuration file when it is parsed from disk. This has previously
been impossible because the individual settings have been spread
across modules and files.

Settings continue to live in separate files but the classes in which
they live are loaded via ServiceLoading. This allows the Config object
to have access to all settings at runtime.

To make sure that the Config object is used a sensible way its
constructors have been made private. The only way it build a Config is
via its static factory methods. This was already the primary way of
constructing it previously.

The new structure allows for two clear entry points:

* `embeddedDefaults()`
* `serverDefaults()`

The names should be self-explanatory.

A secondary effect was it was necessary to be able to pass a Config
object directly to the PlatformModule/GraphDatabase factories instead
of doing a Config to Map conversion.

Converting a Config object to a Map is still possible via the `getRaw()` method,
but should only be done when absolutely required. To
compensate, convenience functions have been added to allow safe and
reasonable access to the raw settings. Primary intended usage remains
passing a Settings object which does its own parsing.

A new maven module `community/configuration` was added to allow
modules to have loadable settings without adding a dependency on
Kernel.

changelog: Improved parsing of configuration

- added a significant amount of config validation to the startup process of neo4j. Any errors will prevent neo4j from starting but they will also have very helpful error messages attached to them.
- added a new config option `dbms.config.strict_validation` which enables even stricter validation (false by default, will be true by default in 4.0).
- added new methods to embedded graph database builder APIs which take a Config object instead of a StringMap.
- deprecated several database builder API methods in favor of the new Config methods.

spacecowboy committed:
Add new configuration module to provide service loadable config

Necessary to allow all settings classes implement the interface to be
service loadable.

Setting implements new interfaces used in the service loading.

spacecowboy committed:
Refactor (Facade)Factory and Platform to use Config not Map

spacecowboy committed:
Make primary use of ConfigLoader static

spacecowboy committed:
Add tests for log passing in config

spacecowboy committed:
Make all settings service loadable

spacecowboy committed:
Refactor Config to use service loading to get available settings

spacecowboy committed:
Fix ConfigTests and related

spacecowboy committed:
Fix log tests and code for nonnull requirements

spacecowboy committed:
Update tests and fix tests for new  config system

spacecowboy committed:
Fix nullable annotation/tests in logging

spacecowboy committed:
Fix tests for config changes

spacecowboy committed:
Add connector validators and new connector utility methods

spacecowboy committed:
Fix tests and add ServerConfigurationValidator with serverDefaults

spacecowboy committed:
Fix another nonnull compilation issue

spacecowboy committed:
Fix missing harness choice of https port

spacecowboy committed:
Fix compilation and update Config use in Community

spacecowboy committed:
Add HA and CC configuration validation

spacecowboy committed:
Fix compilation issues and update Config uses in enterprise

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -54,57 +72,158 @@
+    public static Config embeddedDefaults( Map<String,String> additionalConfig )
+    {
+        return embeddedDefaults( Optional.empty(), additionalConfig );
+    }
```
Would it make sense to also have an `embeddedTestingDefaults` that tests could use? (Ours and our customers tests alike)

spacecowboy replied:
It certainly would but I'm not sure what values that would entail except pagecache size? I'd be happy to take suggestions on the matter

chrisvest replied:
The page cache size is the main differentiator, I think.

spacecowboy committed:
Fix compilation in Desktop

spacecowboy committed:
Restore public API classes but add deprecations on them

spacecowboy committed:
Fix configuration bean

spacecowboy committed:
Fix integration tests

spacecowboy committed:
Fix compilation errors in Desktop

spacecowboy committed:
Fix javadoc generation

spacecowboy committed:
Do not assert at runtime if field is not static

spacecowboy committed:
Fix propertyExistenceConstraintsIT

spacecowboy committed:
Add a file with dummy settings (used in wrapper scripts)

benbc commented on community/configuration/src/main/java/org/neo4j/configuration/ExternalSettings.java:
```diff
@@ -0,0 +1,101 @@
+ * Be aware that values are still validated.
+ */
+@SuppressWarnings( "unused" )
+public class ExternalSettings implements LoadableConfig
```
Could this be provided by the `server` module or something else higher up the stack? `command-line`?

benbc commented on community/dbms/src/main/java/org/neo4j/server/configuration/ConfigLoader.java:
```diff
@@ -21,53 +21,62 @@
-    private final Function<Map<String, String> ,Iterable<Class<?>>> settingsClassesSupplier;
+    public ConfigLoader()
+    {
+    }
```
Why do we need a public constructor as well as the static factory methods? If we can't hid the constructor for some reason then this empty no arg public constructor is not necessary.

benbc commented on community/import-tool/src/main/java/org/neo4j/tooling/ImportTool.java:
```diff
@@ -528,7 +528,7 @@ public static void doImport( PrintStream out, PrintStream err, File storeDir, Fi
     private static Config loadDbConfig( File file ) throws IOException
     {
-        return file != null && file.exists() ? new Config( MapUtil.load( file ) ) : Config.defaults();
+        return file != null && file.exists() ? Config.embeddedDefaults( MapUtil.load( file ) ) : Config.defaults();
```
Should this use the disabled-connectors version just on principle?

spacecowboy replied:
Would be nice but that would introduce a cyclic dependency since ConfigLoader is in dbms, and dbms already depends on importtool.

chrisvest replied:
Could that cycle be solved by moving `ConfigLoader` to `configuration`?

benbc commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseFactory.java:
```diff
@@ -95,10 +96,11 @@ protected void configure( GraphDatabaseBuilder builder )
 
-    protected GraphDatabaseService newDatabase( File storeDir, Map<String,String> config, GraphDatabaseFacadeFactory.Dependencies dependencies )
+    protected GraphDatabaseService newDatabase( File storeDir, Map<String,String> settings,
+            GraphDatabaseFacadeFactory.Dependencies dependencies )
```
Does it make sense to deprecate this method that takes settings as a map?

benbc commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -464,6 +501,31 @@
+
+    // Needed to validate config, accessed via reflection
+    @SuppressWarnings( "unused" )
+    public static final HttpConnectorValidator httpValidator = new HttpConnectorValidator();
```
Gah! I'm deeply traumatised by defining the HTTP config in the kernel. No chance of pushing this up to the server, I suppose?

spacecowboy replied:
I would like to pull all connectors out of the kernel, but bolt being a kernel extension kind of prevents that. There are a lot of stuff in the kernel which shouldn't logically belong there but them being there now forces stuff to be there because of service loading

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -636,12 +637,15 @@ private TransactionLogModule buildTransactionLogs(
+        if ( config.get( GraphDatabaseFacadeFactory.Configuration.ephemeral ) )
+        {
+            config = config.withDefaults(
+                    stringMap( GraphDatabaseSettings.keep_logical_logs.name(), "1 files" ) );
```
This seems an odd place to define this default. I don't understand why it's necessary.

spacecowboy replied:
I didn't write it, I just copied old functionality.

benbc replied:
Github's diff shows that `"1 files"` is new. Where did it come from?

spacecowboy replied:
This piece of code is the logical (and readable) equivalent of the code which it replaced:

```java
        String pruningConf = config.get(
                 config.get( GraphDatabaseFacadeFactory.Configuration.ephemeral )
                 ? GraphDatabaseFacadeFactory.Configuration.ephemeral_keep_logical_logs
                 : GraphDatabaseSettings.keep_logical_logs );
```

That piece of code referenced a piece of duplicated code (which you also commented on) in `GraphDatabaseFacadeFactory` which was:

```java
ephemeral_keep_logical_logs = setting( "dbms.tx_log.rotation.retention_policy", STRING, "1 " + "files", ....)
```

So the code boils down to:

if database is ephemeral
then default for keep_logical_logs is "1 file"
else the regular default defined in GraphDatabaseSettings

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/BoltConnectorValidator.java:
```diff
@@ -0,0 +1,160 @@
+    {
+        final HashMap<String,String> result = new HashMap<>();
+
+        ownedEntries( rawConfig ).forEach( s ->
```
This seems to be basically the same as `values()`, can it not use the same implementation?

spacecowboy replied:
No, `values` calls `values()` (logically enough) on the settings. Whereas `validate` calls `validate()` on the settings.

The result of `values` a map of setting-key, setting-value-object. The values have been converted from the raw string form into objects, like `ListendAddress` for example.

The result of 'validate' on the other hand is a map of the raw key, raw value. The result is the sub-map which this validator considers valid.

benbc replied:
Right -- but can't the structural duplication be removed?

spacecowboy replied:
Not without adding a great deal of obfuscation in my opinion, which I consider much worse

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Connector.java:
```diff
@@ -0,0 +1,64 @@
+
+    // Note: We no longer use the typeDefault parameter because it made for confusing behaviour;
+    // connectors with unspecified would override settings of other, unrelated connectors.
+    // However, we cannot remove the parameter at this
```
Incomplete sentence.

benbc replied:
Should this constructor be deprecated in favour of one without the `typeDefault`?

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/HttpConnectorValidator.java:
```diff
@@ -0,0 +1,224 @@
+import static org.neo4j.kernel.configuration.Settings.options;
+import static org.neo4j.kernel.configuration.Settings.setting;
+
+public class HttpConnectorValidator extends ConnectorValidator
```
Is it possible to move the bits of `validate()/values()` that deal with common connector stuff up into `ConnectorValidator` so as to remove duplicate with `BoltConnectorValidator`?

spacecowboy replied:
I believe that would make things hard to read for the very marginal benefit of moving `enabled` out of the switch clause. `type` is a shared setting but each validator asserts that it is indeed correct in terms of itself (http vs bolt). It would also mean we'd need to remove the `default` case where unknown settings cause errors (signs of programmer error).

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/HttpConnectorValidator.java:
```diff
@@ -0,0 +1,224 @@
+    @Nonnull
+    public Map<String,String> validate( @Nonnull Map<String,String> rawConfig )
+            throws InvalidSettingException
+    {
```
Same as `BoltConnectorValidator`, there is duplication with `values()`.

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -83,15 +85,10 @@
-                setting( "dbms.tx_log.rotation.retention_policy", STRING, "1 " + "files", illegalValueMessage( "must be `true`/`false` or of format '<number><optional unit> <type>' " +
-                        "for example `100M size` for " +
-                        "limiting logical log space on disk to 100Mb," +
-                        " or `200k txs` for limiting the number of transactions to keep to 200 000", matches( ANY ) ) );
```
Why is it okay to remove this?

spacecowboy replied:
It was a duplicate of the same setting present in `GraphDatabaseSettings.java`

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/internal/EmbeddedGraphDatabase.java:
```diff
@@ -72,6 +73,16 @@ public EmbeddedGraphDatabase( File storeDir,
+     */
+    public EmbeddedGraphDatabase( File storeDir,
+            Config config,
+            GraphDatabaseFacadeFactory.Dependencies dependencies )
```
Where is this called from?

spacecowboy replied:
`ImpermanentGraphDatabase`, used in testing. Was able to make it protected.

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/ServerConfigurationValidator.java:
```diff
@@ -0,0 +1,90 @@
+import static org.neo4j.kernel.configuration.Connector.ConnectorType.HTTP;
+
+
+public class ServerConfigurationValidator implements ConfigurationValidator
```
Does this really need to live down in the kernel?

spacecowboy replied:
Yes, because the Config class has a `serverDefaults` method which requires this validator.

benbc replied:
Could that method be generalized into something that would allow the actual server-specific stuff to move out of Kernel? I'm worried that this and the move of the HTTP stuff down to sit alongside the Bolt stuff is making things worse in a way that we don't have a plan to improve.

benbc commented on community/server/src/main/java/org/neo4j/server/CommunityNeoServer.java:
```diff
@@ -54,7 +54,7 @@
         File storeDir = config.get( DatabaseManagementSystemSettings.database_path );
         return new GraphDatabaseFacadeFactory( DatabaseInfo.COMMUNITY, CommunityEditionModule::new )
-                .newFacade( storeDir, config.getParams(), dependencies );
+                .newFacade( storeDir, config, dependencies );
```
Hurray!

benbc commented on enterprise/backup/src/main/java/org/neo4j/backup/BackupService.java:
```diff
@@ -217,11 +219,20 @@ private BackupOutcome incrementalBackup( FileSystemAbstraction fileSystem, Strin
+        {
+            Optional<String> value = config.getRaw( key );
+            value.ifPresent( s -> configParams.put( key, s ) );
+        }
```
What's this doing?

spacecowboy replied:
This piece of logic is converting the Config object into a StringMap. Required because GraphDatabaseBuilder does not yet accept a Config object (see comment above about required breaking interface changes)

chrisvest replied:
Can't such support be added without breaking existing API?

benbc commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/CoreGraphDatabase.java:
```diff
@@ -47,13 +48,28 @@ public CoreGraphDatabase( File storeDir, Map<String,String> params,
+    }
+
     public CoreGraphDatabase( File storeDir, Map<String,String> params,
             GraphDatabaseFacadeFactory.Dependencies dependencies, DiscoveryServiceFactory discoveryServiceFactory )
```
What is this still used for?

benbc commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/readreplica/ReadReplicaGraphDatabase.java:
```diff
@@ -44,12 +45,26 @@ public ReadReplicaGraphDatabase( File storeDir, Map<String,String> params, Depen
+    }
+
     public ReadReplicaGraphDatabase( File storeDir, Map<String,String> params, Dependencies dependencies,
             DiscoveryServiceFactory discoveryServiceFactory )
```
What is this still used for (as above)?

benbc commented on enterprise/cluster/src/main/java/org/neo4j/cluster/ClusterSettings.java:
```diff
@@ -69,8 +71,22 @@ public String toString()
+    @Description( "Configure the operating mode of the database -- 'SINGLE' for stand-alone operation, 'HA' for " +
+            "operating as a member in a cluster, 'ARBITER' for an HA-only cluster member with no database, " +
+            "CORE for a core member of a Causal Clustering cluster, or READ_REPLICA for read replica." )
+    public static final Setting<Mode> mode = setting( "dbms.mode", options( Mode.class ), Mode.SINGLE.name() );
```
This seems like a strange place to define this (since it crosses HA and core-edge. Where was it before?

spacecowboy replied:
[EnterpriseServerSettings](https://github.com/neo4j/neo4j/blob/ddd4cbac3db01f0994fa3f5eefc656af59879861/enterprise/server-enterprise/src/main/java/org/neo4j/server/enterprise/EnterpriseServerSettings.java#L39)

which is a module higher than both HA and CC in the hierarchy (and it must reside in a module which precedes them both).

It's not the ideal location but I'm not sure such a place exists at the moment and I didn't want to create yet another maven module for this single setting.

benbc replied:
Sorry -- why can't it still be in `EnterpriseServerSettings`?

spacecowboy replied:
because `EnterpriseServer` depends on HA and CC, not the other way round

benbc reviewed (commented):
Have you considered plugin-writers? I believe that some plugins have config which is written into `neo4j.conf`. How will this work for them? Maybe discus with Stefan Armbruster.

What is the experience like for Neo4j developers adding new settings classes (say a new module)? How will they discover that they have forgotten the service loader stuff?

benbc commented:
`Config.defaults()`: is this helpful? Why not force people to choose between `serverDefaults()` and `embeddedDefaults()`?

benbc commented:
Can `Config.serverDefaults()` use `withDefaults()` instead of doing its own `putIfAbsent` stuff?

benbc commented:
`Config.replaceSettings()`: can the extra validator (`IndividualSettingsValidator`) just be added to the list of validators locally to avoid the call to `.validate()` outside the loop?

spacecowboy committed:
Remove unused constructor in ConfigLoader

spacecowboy committed:
Deprecate factory method which accepts a map

spacecowboy committed:
Deprecate Connector constructor with typeDefault

spacecowboy committed:
Make internal embedded constructor protected

spacecowboy committed:
Remove unnecessary cluster database constructors

spacecowboy commented:
@benbc in answer to your questions

> Have you considered plugin-writers? I believe that some plugins have config which is written into neo4j.conf. How will this work for them? Maybe discus with Stefan Armbruster.

If they want to put stuff in the neo4j config they'll have to have a settings class which is service loadable too. Our docs on plugins seem to be almost non-existing for the 3.x branches. I'm not sure what our stance on this is.

> What is the experience like for Neo4j developers adding new settings classes (say a new module)? How will they discover that they have forgotten the service loader stuff?

They'll get warnings in the log about "Unknown settings".

> Config.defaults(): is this helpful? Why not force people to choose between serverDefaults() and embeddedDefaults()?

I agree it should be done but it would grow the diff a lot, making reviewing even harder. It can be done once other changes are final.

> Can Config.serverDefaults() use withDefaults() instead of doing its own putIfAbsent stuff?

No because the validator would be executed and throw an exception about missing http connector before you get to call `withDefaults()`.

> Config.replaceSettings(): can the extra validator (IndividualSettingsValidator) just be added to the list of validators locally to avoid the call to .validate() outside the loop?

Not in a straight-forward way. Validators are passed on to child config objects when `with` methods are called. Adding the individualValidator to the list would introduce another duplicate every time a `with` method was called.

Pushed an update with changes to address other comments.

benbc commented:
> > Have you considered plugin-writers? I believe that some plugins have config which is written into neo4j.conf. How will this work for them? Maybe discus with Stefan Armbruster.

> If they want to put stuff in the neo4j config they'll have to have a settings class which is service loadable too. Our docs on plugins seem to be almost non-existing for the 3.x branches. I'm not sure what our stance on this is.

Please do talk to a human about what people actually do in practice, rather than relying on the lack of docs to consider the behaviour unspecified. This is the kind of thing that causes pain for Field and makes them angry with Engineering when we get it wrong.

benbc commented:
> > What is the experience like for Neo4j developers adding new settings classes (say a new module)? How will they discover that they have forgotten the service loader stuff?

> They'll get warnings in the log about "Unknown settings".

That's very optimistic of you. :-).

Do you think we should consider a breaking change for 4.0 where you can no longer use settings which haven't been defined?

spacecowboy committed:
Only warn and discard unknown settings in our own namespace

chrisvest commented on community/configuration/src/test/java/org/neo4j/configuration/LoadableConfigTest.java:
```diff
@@ -0,0 +1,176 @@
+        assertEquals( "A string setting", options.get( 1 ).description().get() );
+    }
+
+    private static class Test implements LoadableConfig
```
You could name this `TestConfig` and avoid the conflict with the `org.junit.Test` class.

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseBuilder.java:
```diff
@@ -39,8 +39,13 @@
+         * This method will no longer accept a map in 4.0, but a Config object instead.
+         */
+        @Deprecated
         GraphDatabaseService newDatabase( Map<String, String> config );
```
You could already now add the `newDatabase( Config )` method, and turn this old declaration into a `default` method implemented as `newDatabase( Config.empty().augment( config ) );`. Or so I think.

spacecowboy replied:
Not exactly. That would break the implementation of anyone implementing the interface. To be non-breaking the new method would have to have a default implementation which delegates to the old method. But good idea nonetheless! See changes.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -153,28 +271,83 @@ public Config with( Map<String, String> additionalConfig, Class<?>... settingsCl
+     * Replays possible existing messages into this log
+     *
+     * @param log to use
+     */
```
The main function of this method is assigning the `log` field. The replying of log messages is a side-effect, if the previous value was a `BufferingLog`.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/ConnectorValidator.java:
```diff
@@ -0,0 +1,127 @@
+import static org.neo4j.kernel.configuration.Settings.setting;
+
+
+public abstract class ConnectorValidator implements SettingGroup<Object>
```
Two consecutive empty lines might make checkstyle complain.

chrisvest commented on enterprise/server-enterprise/src/main/java/org/neo4j/server/enterprise/EnterpriseServerSettings.java:
```diff
@@ -29,29 +30,22 @@
-            "CORE for a core member of a Causal Clustering cluster, or READ_REPLICA for read replica." +
-            " cluster" +
-            ". " )
-    Setting<String> mode = setting( "dbms.mode", STRING, "SINGLE" );
```
Why is it okay to remove this?

spacecowboy replied:
It has been moved (and converted to an enum) to [ClusterSettings.java](https://github.com/neo4j/neo4j/blob/c8932e8f5e8b1eda8f905bb8b33cc8059a16121e/enterprise/cluster/src/main/java/org/neo4j/cluster/ClusterSettings.java)

reason being that the mode setting must be accessible by all the cluster database creators, as well as the server endpoint. The cluster module satisfied that requirement, though it is not the perfect place for this setting, I agree.

spacecowboy committed:
Avoid naming conflict with junit.Test

spacecowboy committed:
Add new non-deprecated factory method to DatabaseCreator

with a default implementation so external implementors don't
suffer. Implement it everywhere internally however

spacecowboy committed:
Fix some style issues

spacecowboy commented:
thanks for you feedback @chrisvest , I have updated accordingly.

spacecowboy committed:
Update license headers for 2017

spacecowboy committed:
Fix merge conficts

spacecowboy committed:
Added new config option: dbms.config.strict_validation

By default this setting is false, for now. It causes config parsing to
throw errors on unknown settings (in our namespace).

Settings outside our namespaces are still retained and not warned
about.

Also changed so that warning is a flag, and it is only true when
reading the config file (so just once).

spacecowboy committed:
Reduce duplication in connector validators and make them more strict

benbc commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -110,6 +107,11 @@
+    @Description( "A strict configuration validation will prevent the database from starting up if unknown " +
+            "configuration options are specified in the neo4j settings namespace (such as dbms., ha., cypher., etc)." )
+    public static final Setting<Boolean> strict_config_validation =
+            setting("dbms.config.strict_validation", BOOLEAN, FALSE );
```
Maybe put a note that the default will switch to `true` in a future version.

spacecowboy committed:
Fix broken tests

spacecowboy committed:
Merge remote-tracking branch 'neo4j/3.2' into 3.2-configstuff

spacecowboy committed:
Merge branch '3.2' of github.com:neo4j/neo4j into 3.2-configstuff
